Adding an implementation for timerfd.

vdagonneau-anssi opened the pull request:
Hello,

I have encountered a situation where I needed to use timerfd in addition to other fd with nix implementation of epoll. 

Here is a basic implementation of this feature, I also added some documentation and tests.

Note: currently testing works by pausing for 1-3 seconds. Obviously it is pretty bad if we want tests to run quickly but that was my best idea at the time. Feel free to suggest any other way.

Thank you!

asomers-ax commented on test/sys/test_timerfd.rs:
```diff
@@ -0,0 +1,49 @@
+    read(fd.as_raw_fd(), &mut buffer).unwrap();
+
+    let millis = before.elapsed().as_millis();
+    assert!(millis < 1100 && millis > 900);
```
You cannot assert that `millis < 1100`.  There is no guarantee that the timer will expire promptly, or that this process will run immediately after the timer expires.

vdagonneau-anssi replied:
You are right, I removed the upper bound. Do you think setting a lower bound could be problematic in some corner cases ? I could not think of any "practical" problems with that.

asomers replied:
I don't know of any case where the lower bound could be a problem, unless the read gets interrupted by a signal.  What does Linux do in that case?  If it's possible for the read to return early with EINTR, then you should retry the read in your higher-level method.

asomers-ax commented on test/sys/test_timerfd.rs:
```diff
@@ -0,0 +1,49 @@
+    read(fd.as_raw_fd(), &mut buffer).unwrap();
+
+    let start_delay = before.elapsed().as_millis();
+    assert!(start_delay < 1100 && start_delay > 900);
```
Ditto as above.  You can't assert an upper bound for the delay.

vdagonneau-anssi replied:
Same as above.

asomers-ax commented on src/sys/timerfd.rs:
````diff
@@ -0,0 +1,124 @@
+//! # Examples
+//!
+//! Create a new one-shot timer that expires after 5 seconds.
+//! ```no_run
````
Why the `no_run`?  You might as well run it, though maybe decrease the timer to 1 second.  Also, the example would be more useful if there were a `read` at the end.

vdagonneau-anssi replied:
Ok I removed the no_run, decreased the timer and added a read at the end.

asomers-ax commented on src/sys/timerfd.rs:
```diff
@@ -0,0 +1,124 @@
+
+impl TimerFd {
+    pub fn new(clockid: ClockId, flags: TimerFlags) -> Result<Self> {
+        let clid = match clockid {
```
I think you can skip this block if you define ClockId using the `libc_enum` macro.

vdagonneau-anssi replied:
Ok refactored ClockId to use libc_enum.

asomers-ax commented on src/sys/timerfd.rs:
```diff
@@ -0,0 +1,124 @@
+        Errno::result(unsafe { libc::timerfd_create(clid, flags.bits()) }).map(|fd| Self { fd })
+    }
+
+    pub fn settime(&self, flags: TimerSetTimeFlags, expiration: Expiration) -> Result<()> {
```
Could we have docs for all public functions?

asomers-ax commented on src/sys/timerfd.rs:
```diff
@@ -0,0 +1,124 @@
+//! Timer FD is a Linux-only API to create timers and get expiration notifications through file
+//! descriptors.
+//!
+//! For more documentation, please read [timerfd_create(2)](http://man7.org/linux/man-pages/man2/timerfd_create.2.html).
```
Wrap to 80 cols, please.

vdagonneau-anssi replied:
Ok, wrapped everything to 80 cols.

asomers-ax commented on test/sys/test_timerfd.rs:
```diff
@@ -0,0 +1,49 @@
+
+#[test]
+pub fn test_timerfd_oneshot() {
+    let mut buffer = [0u8; 64];
```
If buffer is never read, can it be a 0-length slice?

vdagonneau-anssi replied:
So, no I cannot make it a 0-length slice, the syscall requires a buffer that is no less that 8-bytes long. If you don't you get an EINVAL.

That said 64-bytes is way too big. I changed it to use the minimum required size of 8-bytes.

vdagonneau-anssi commented:
I addressed most of your comments in the latest commit I think. 

As for having a sleep method or equivalent, I think it is a fairly `high level` interface and I will need to think about how to implement it properly.

It is worth noting that reading on the `fd` has different meanings depending on how you created the timerfd in the first place (or modified it with fcntl). Specifically, the call to read can either be blocking or not.

If we indeed need a higher level interface then I would like to move away from the method names I have right now to make it clear that the semantics are different from the underlying syscalls.

asomers commented on src/sys/timerfd.rs:
```diff
@@ -1,33 +1,45 @@
 //! // We set a new one-shot timer in 5 seconds.
-//! fd.settime(TimerSetTimeFlags::empty(), Expiration::OneShot(TimeSpec::seconds(5))).unwrap();
+//! fd.settime(
+//!     Expiration::OneShot(TimeSpec::seconds(5)),
```
Be sure to set the time to just 1 second.  5 seconds is too long for a doc test.

vdagonneau-anssi replied:
Ok, fixed it.

asomers commented on src/sys/timerfd.rs:
```diff
@@ -1,33 +1,45 @@
+use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
 
+/// A timerfd instance. This is also a file descriptor, you can feed it to
+/// other interfaces consuming file descriptors, epoll for example.
```
Ahh, nice addition here.  I didn't realize that timerfd could be used with epoll.  That's probably why it was created, I'm guessing.

vdagonneau-anssi replied:
Yes, that is indeed why it was created.

vdagonneau-anssi commented:
I pushed a new set of changes including a slightly higher level interface.

You now have 5 methods for the TimerFd: `new`, `set`, `get`, `unset`and `wait`.

The idea is that you use `new` to create a new timerfd, it is created without any expiration. You can then `set` a new expiration and `get` it back. If you wish you can also `unset` any  previous expiration that has been set. Finally, you can "acknowledge"/"wait" the timer with `wait`.

I still feel somewhat uneasy about the `wait` name as it really doesn't make a lot of sense in the context of `epoll`.

Typically in the case of `epoll`, you would create a timerfd, set an expiration, feed it to epoll. Once the timer is triggered, epoll will notify you that the timerfd has become readable and you need to "acknowledge" it (a.k.a. read from it) otherwise epoll is going to keep notifying you about it.

The name `wait` works well in a conventional approach where you create the timerfd, set an expiration and the read from it. In that case `wait` will indeed wait until timerfd becomes readable.

vdagonneau-anssi commented:
Ok, I added it to the changelog.

asomers commented:
Only one more problem: you'll have to rebase to fix the test failure.

vdagonneau-anssi commented:
Ok, I rebased on top of master.
